### PR TITLE
Portia ORM on Django Fixes

### PR DESCRIPTION
**Pull request by ruairif**
**Add FS backend support**
**Remove old ORM and use portia_server ORM in slyd websocket**
**Backport changes from nui-develop:**
- Add generated start urls support to ORM
- Add repeated fields extraction

**Fix ORM compatabilities with old/broken projects:**
- Add missing fields references by annotations
- If an annotation doesn't have a field create a new one
- Add missing schemas references by annotations
- Load all sample data when making `GET /api/projects/<PID>/spiders/<SID>/samples` request
- Add body field to samples
- Repair truncated ids for annotations
- Create deterministic id generator for fields
- Disable DRF `format` url parameter
- Remove non existent extractors from annotations

**Fix `wrapped C/C++ object of type SplashQWebPage has been deleted` error**
**Fix repeated fields extraction**

### CODE REVIEW

File: slyd/slyd/utils/storage.py [hunk 1]
```diff
@@ -1,34 +1,114 @@
...
+from .repoman import Repoman, DEFAULT_USER, FILE_MODE
 
 
-FILE_MODE = 0o100644
+class InvalidFilename(Exception):
+    pass
 
 
-class CommitingStorage(object):
+class CommittingStorage(object):
```
Comment (sagelliv) [inline slyd/slyd/utils/storage.py]: I think this object is not getting us any benefit here. Can we move these methods to `BasePortiaStorage` as all storage classes will inherit from this?

File: slyd/slyd/utils/storage.py [hunk 2]
```diff
@@ -42,31 +122,87 @@ def move(self, old_file_name, new_file_name, allow_overwrite=False):
     def rmtree(self, name):
         shutil.rmtree(self.path(name))
 
+    def _save(self, name, content):
```
Comment (sagelliv) [inline slyd/slyd/utils/storage.py]: This is being duplicated in `portia_server`. Can we import classes from that directory or is there some reason to keep these separate?

File: slyd/slyd/jsonapi/serializers.py [hunk 3]
```diff
@@ -100,7 +101,7 @@ def __new__(mcs, name, bases, attrs):
         cls = super(JsonApiSerializerMeta, mcs).__new__(mcs, name, bases, attrs)
 
         # add new schema to registry by type
-        serializers[schema_type] = cls
+        schemas[schema_type] = cls
```
Comment (sagelliv) [inline slyd/slyd/jsonapi/serializers.py]: Should be serializers (changing the comment as well). Marshmallow has the notion of schemas, but we are really using them as serializers (as portrayed by the meta class name `JsonApiSerializerMeta`.

File: slyd/slyd/jsonapi/serializers.py [hunk 4]
```diff
@@ -419,7 +420,7 @@ def load_profile_updates(self):
             }
 
             try:
-                serializer_class = get_serializer(type_)
+                serializer_class = get_schema(type_)
```
Comment (sagelliv) [inline slyd/slyd/jsonapi/serializers.py]: `get_serializer`

File: slyd/slyd/jsonapi/serializers.py [hunk 5]
```diff
@@ -548,7 +549,7 @@ def add_includes(self, includes):
         included_data = self.included_data
         for instance in includes:
             type_ = type_from_model_name(instance.__class__.__name__)
-            serializer = get_serializer(type_)(
+            serializer = get_schema(type_)(
```
Comment (sagelliv) [inline slyd/slyd/jsonapi/serializers.py]: get_serializer

File: slyd/slyd/jsonapi/serializers.py [hunk 6]
```diff
@@ -561,7 +562,7 @@ def format_profile_references(self, instances):
         references = []
         for instance in instances:
             type_ = type_from_model_name(instance.__class__.__name__)
-            serializer = get_serializer(type_)(
+            serializer = get_schema(type_)(
```
Comment (sagelliv) [inline slyd/slyd/jsonapi/serializers.py]: get_serializer

File: slyd/slyd/resources/route.py [hunk 7]
```diff
@@ -12,12 +12,12 @@
...
                        JsonApiNotFoundResponse, JsonApiValidationErrorResponse)
 from ..errors import BadRequest, BaseError
-from ..jsonapi.serializers import JsonApiPolymorphicSerializer
-from ..jsonapi.registry import get_serializer
-from ..jsonapi.utils import type_from_model_name
-from ..orm.collection import ModelCollection
-from ..orm.exceptions import ProtectedError
-from ..orm.relationships import BelongsTo, HasMany
+from portia_api.jsonapi.serializers import JsonApiPolymorphicSerializer
+from portia_api.jsonapi.registry import get_schema
```
Comment (sagelliv) [inline slyd/slyd/resources/route.py]: get_serializer

File: slyd/slyd/resources/route.py [hunk 8]
```diff
@@ -170,8 +170,8 @@ def get_serializer(self, instance=None, data=None, many=False, **kwargs):
                 instance=instance, data=data, many=many, **params)
 
         type_ = type_from_model_name(self.default_model.__name__)
-        return get_serializer(type_)(
-            instance=instance, data=data, many=many, **params)
+        return get_schema(type_)(instance=instance, data=data, many=many,
```
Comment (sagelliv) [inline slyd/slyd/resources/route.py]: get_serializer

**Comment by sagelliv** [comment]
Fix `ImportError` for the `slyd` tests, but even better we should move the ORM tests to django and call them from there during CI.

File: portia_server/portia_orm/models.py [hunk 9]
```diff
@@ -346,27 +349,38 @@ def load(cls, storage, instance=None, spider=None, **kwargs):
...
+            return data
         methods = (self.migrate_sample, self.get_items)
         for method in methods:
             data = method(self, data)
         return data
 
     @staticmethod
     def migrate_sample(self, data):
+        if not data.get('body'):
+            data['body'] = 'original_body'
```
Comment (ruairif) [inline portia_server/portia_orm/models.py]: This has been changed to:
```python
data['body'] = data.get('body') or 'original_body'
```

File: slyd/slyd/utils/storage.py (hunk 1 above)
Comment (ruairif) [inline slyd/slyd/utils/storage.py]: I have left `CommittingStorage` as it is but I have restructured the `BasePortiaStorage`
Now `CommitingStorage` contains methods used only by `Portia` and `BasePortiaStorage` overrides methods that reside in `django.core.files.storage.Storage`.
This separation better maps onto the default use case within django

File: slyd/slyd/utils/storage.py (hunk 2 above)
Comment (ruairif) [inline slyd/slyd/utils/storage.py]: This file has been deleted I think

File: slyd/slyd/jsonapi/serializers.py (hunk 4 above)
Comment (ruairif) [inline slyd/slyd/jsonapi/serializers.py]: done

File: slyd/slyd/jsonapi/serializers.py (hunk 6 above)
Comment (ruairif) [inline slyd/slyd/jsonapi/serializers.py]: done

File: slyd/slyd/resources/route.py (hunk 7 above)
Comment (ruairif) [inline slyd/slyd/resources/route.py]: done

File: slyd/slyd/resources/route.py (hunk 8 above)
Comment (ruairif) [inline slyd/slyd/resources/route.py]: done

File: slyd/slyd/jsonapi/serializers.py (hunk 5 above)
Comment (ruairif) [inline slyd/slyd/jsonapi/serializers.py]: done

File: slyd/slyd/jsonapi/serializers.py (hunk 3 above)
Comment (ruairif) [inline slyd/slyd/jsonapi/serializers.py]: done

**Pull request merged**